value.go

```diff
@@ -357,24 +357,29 @@ func args(xs ...*cachedThunk) callArguments {
 
 // Object is a value that allows indexing (taking a value of a field)
 // and combining through mixin inheritence (operator +).
-//
-// Note that every time a field is indexed it evaluates it again, there is
-// no caching of field values. See: https://github.com/google/go-jsonnet/issues/113
-type valueObject interface {
-	value
-	inheritanceSize() int
-	index(i *interpreter, trace TraceElement, field string) (value, error)
-	assertionsChecked() bool
-	setAssertionsCheckResult(err error)
-	getAssertionsCheckResult() error
+
+type valueObject struct {
+	valueBase
+	assertionError error
+	cache          map[objectCacheKey]value
+	underlying     objectBody
```

Comment:
Any ideas for a better names than "underlying" and "objectBody"?

Reply:
We could rename the top-level one valueCachedObject and then the other one could just be object or valueObject or uncachedObject.  I don't really like the word "body" as it usually implies the bit that isn't a header but there's no header here.

Reply:
The `objectBody`, so valueObject is out of the question. I like the `uncachedObject` name the most.

So I renamed `objectBody` to `uncachedObject` and `underlying` to `uncached`. I left `valueObject` as it was, because I think it's clear enough - it is the only representation of object values (uncachedObject is not a value on its own).